packages/react-checkbox/src/components/Checkbox/useCheckbox.tsx

```diff
@@ -28,45 +26,57 @@ const mergeProps = makeMergeProps<CheckboxState>({ deepMerge: checkboxShorthandP
  * @param props - props from this instance of Checkbox
  * @param ref - reference to root HTMLElement of Checkbox
  */
-export const useCheckbox = (props: CheckboxProps, ref: React.Ref<HTMLElement>): CheckboxState => {
-  const state: CheckboxState = mergeProps(
-    {
-      ref,
-      as: props.children !== undefined ? Label : 'span',
-      id: useId('checkbox-'),
-      size: 'medium',
-      labelPosition: 'after',
-      input: {
-        as: 'input',
-        type: 'checkbox',
-        children: null,
-      },
-      indicator: {
-        as: 'div',
-      },
-    },
-    resolveShorthandProps(props, checkboxShorthandProps),
-  );
-
+export const useCheckbox = (
+  {
+    disabled = false,
+    circular = false,
+    required = false,
+    id,
+    defaultChecked,
+    checked: propsChecked,
+    ...props
+  }: CheckboxProps,
+  ref: React.Ref<HTMLElement>,
+): CheckboxState => {
   const [checked, setCheckedInternal] = useControllableState({
-    defaultState: props.defaultChecked,
-    state: props.checked,
+    defaultState: defaultChecked,
+    state: propsChecked,
     initialState: false,
   });
 
+  const state: CheckboxState = {
+    circular,
+    checked: checked,
+    size: 'medium',
+    labelPosition: 'after',
+    components: { root: props.children !== undefined ? Label : 'span', indicator: 'div', input: 'input' },
+    input: resolveShorthand(props.input, {
+      required: true,
+      defaultProps: {
+        disabled,
+        type: 'checkbox',
+        required,
+      },
+    }),
+    indicator: resolveShorthand(props.indicator, {
+      required: true,
+    }),
+    root: getNativeElementProps('div', { ref, ...props }),
+  };
+
   const setChecked = React.useCallback(
     (ev: React.ChangeEvent<HTMLInputElement>, val: boolean | 'mixed') => {
-      const onChange = state.onChange;
+      const onChange = props.onChange;
       onChange?.(ev, { checked: val });
```

Comment:
```suggestion
      props.onChange?.(ev, { checked: val });
```

_nit_ This assignment looks redundant...